DisableServiceLinks admission controller

jmcgrath207 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature


<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Adds an optional admission controller to disable EnableServiceLinks for new Pods without the need for a third-party mutating webhook.


#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes # https://github.com/kubernetes/kubernetes/issues/121787

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

Thios PR was reverted in #125002.

The original relnote, for posterity: Added `DisableServiceLinks` admission controller: An optional mutating controller that will set `Spec.EnableServiceLinks: False` for all pods manifests in the cluster before scheduling. 

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

Agreement to solve this issue with an optional admission controller.
REF: https://github.com/kubernetes/kubernetes/issues/121787#issuecomment-1863694095

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

carlory commented:
/ok-to-test

aojea commented on plugin/pkg/admission/disableservicelinks/admission.go:
```diff
@@ -0,0 +1,71 @@
+// newDisableServiceLinks creates a new instance of the DisableServiceLinks admission controller.
+func newDisableServiceLinks() *plugin {
+	return &plugin{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
```
I think this should operate only on Create, IIUIC Updates to a Pod will not recreate the environment variables

jmcgrath207 replied:
Good Catch! 

I did some testing in kind and found those fields are immutable when updating a pod. I removed `admission.Update` in my commit.


```yaml
$ kubectl apply -f simple-pod.yaml 
The Pod "nginx" is invalid: spec: Forbidden: pod updates may not change fields other than `spec.containers[*].image`,`spec.initContainers[*].image`,`spec.activeDeadlineSeconds`,`spec.tolerations` (only additions to existing tolerations),`spec.terminationGracePeriodSeconds` (allow it to be set to 1 if it was previously negative)
  core.PodSpec{
  	... // 26 identical fields
  	RuntimeClassName:          nil,
  	Overhead:                  nil,
- 	EnableServiceLinks:        &true,
+ 	EnableServiceLinks:        &false,
  	TopologySpreadConstraints: nil,
  	OS:                        nil,
  	... // 2 identical fields
  }



```

thockin replied:
Supporting update might be a no-op today, but could be future-safe

aojea commented on plugin/pkg/admission/disableservicelinks/admission.go:
```diff
@@ -0,0 +1,71 @@
+	*admission.Handler
+}
+
+func boolPointer(b bool) *bool {
+	return &b
+}
```
"k8s.io/utils/ptr" has already this tool if you want ptr.To[bool]()

jmcgrath207 replied:
Thanks! I changed my code.

aojea commented on plugin/pkg/admission/disableservicelinks/admission.go:
```diff
@@ -0,0 +1,71 @@
+		return errors.NewBadRequest(fmt.Sprintf("expected *core.Pod but got %T", attributes.GetObject()))
+	}
+
+	pod.Spec.EnableServiceLinks = boolPointer(false)
```
@thockin this will be great if we can differentiate between default and user set , I tried using managed fields but I couldn't see any difference https://github.com/kubernetes/kubernetes/issues/121787#issuecomment-1880029251

thockin replied:
Yeah @jpbetz confirms what I suspected - we just don't have that information. It's unclear how we would want to solve it, and we should not make this PR become the battleground :)

jmcgrath207 commented:
/test pull-kubernetes-e2e-kind-ipv6
/test pull-kubernetes-dependencies

aojea commented on plugin/pkg/admission/disableservicelinks/admission.go:
```diff
@@ -0,0 +1,68 @@
+		return errors.NewBadRequest(fmt.Sprintf("expected *core.Pod but got %T", attributes.GetObject()))
+	}
+
+	pod.Spec.EnableServiceLinks = ptr.To(false)
```
if we don't find a way to differentiate between defauled by the apiserver or set by the user adding Always to the plugin name may be more accurate, `AlwaysDisableServiceLinks`

is just a thought wait for discussion of other reviewers

aojea commented:
one last comment https://github.com/kubernetes/kubernetes/pull/122631#discussion_r1444249514 but LGTM
/assign @thockin

HirazawaUi commented on plugin/pkg/admission/disableservicelinks/admission.go:
```diff
@@ -0,0 +1,68 @@
+	"context"
+	"fmt"
+	"io"
+	"k8s.io/apimachinery/pkg/api/errors"
```
We usually imports are organized in groups, with groups separated by blank lines.

ref: [CodeReviewComments#imports](https://go.dev/wiki/CodeReviewComments#imports)

jmcgrath207 replied:
Thanks! fixed.

HirazawaUi commented on plugin/pkg/admission/disableservicelinks/admission_test.go:
```diff
@@ -0,0 +1,83 @@
+package disableservicelinks
+
+import (
+	"context"
```
Same as above.

jmcgrath207 replied:
fixed

leilajal commented:
/triage accepted

carlory commented:
please add a release note

jmcgrath207 commented:
> please fill `Does this PR introduce a user-facing change?` to add a release note

Fixed

carlory commented:
I think it should have a user-facing change, please add a description about the EnableServiceLinks plugin.

jmcgrath207 commented:
@carlory Ah, I misunderstood. I've updated the release note description. Does this work for you?

thockin commented:
I'll approve and hold, if you want to re-add Update or rename it, ping me.  Else just clear the hold.

/approve
/hold

jmcgrath207 committed:
DisableServiceLinks admission controller

jmcgrath207 commented on plugin/pkg/admission/disableservicelinks/admission_test.go:
```diff
@@ -0,0 +1,101 @@
+		},
+		{
+			description: "Update empty pod with Spec.EnableServiceLinks set to true",
+			requestedPod: core.Pod{
```
An updated operation on a pod does not default to `Spec.EnableServiceLinks=false` but returns a `Spec.EnableServiceLinks=nil` instead.

I am assuming this is expected behavior where defaults are not injected in an updated operation.

jmcgrath207 commented:
Thanks @thockin 

I've added the noop for `admission.Update` and updated the tests to reflect the admission `operation`. I also rebased and squashed. 

If there is nothing else, I am ready to merge.

thockin commented:
Ugh, I missed this ping in the code freeze chaos.

thockin reviewed (commented):
/lgtm
/approve

enj commented:
IMO this PR is the wrong approach, and we should consider reverting it before we release v1.31.  https://github.com/kubernetes/enhancements/issues/3962 should give users the ability to tweak the default as they desire though the Kube API, and we don't need to create one off admission plugins for every wrong default we have (especially since admission config isn't available to users on some environments).

thockin commented:
I don't have any philosophical preference for this sort of laser-focused admission plugins over MAP, except that MAP is alpha.  At the time this proposal was in flgiht I wasn't so aware of MAP.  That said, the interplay between default values and admission control sequencing leaves me scratching my head still.  In a MAP or mutating webhook, I can't tell if a value was specified by a user or by the defaulting logic.  This PR would probably consider that if it could, but it can't, so it doesn't. 

This issue is not so urgent that it can't wait for MAP, though I hope it's not TOOO long.  So I guess I agree with @enj - are there any major reasons why we couldn't back this out and pivot the solution to MAP?

@jmcgrath207 ?

thockin commented:
ping to @jmcgrath207 one more time

jmcgrath207 commented:
Hey All,

Sorry for the late response.

I went ahead and reviewed [KEP-3962](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3962-mutating-admission-policies), and I agree this meets my needs.

I feel it's best to revert my commit and move forward with the KEP-3962 solution to prevent a deprecation scenario. The Alpha part doesn't bother me since both solutions would require a control-plane config change.

As for urgency, users can use a policy controller like kyverno that I mentioned here [here](https://github.com/kubernetes/kubernetes/issues/121787#issuecomment-1804294687) or it can just edit the manifest. I assume I am in the minority here with heavy restrictions on image approval running in their environment.

thockin commented:
Do you want to send a revert or shall I?

jmcgrath207 commented:
@thockin Here is the revert PR. If this isn't the proper way, I have no issue with you reverting it.


https://github.com/kubernetes/kubernetes/pull/125002
